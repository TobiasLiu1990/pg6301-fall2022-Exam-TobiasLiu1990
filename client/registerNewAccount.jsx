import * as React from "react";
import { useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormInput } from "./lib/formInput";

export function RegisterNewAccount() {
  const [role, setRole] = useState("");
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [fullName, setFullName] = useState("");
  const navigate = useNavigate();

  async function handleSubmit(e) {
    e.preventDefault();
    navigate("../");

    await fetch("/api/login/register", {
      method: "post",
      body: JSON.stringify({ username, password, fullName }),
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  function handleSubmitBack() {
    navigate("/");
  }

  return (
    <div>
      <form onSubmit={handleSubmitBack}>
        <button>Go back to start page</button>
      </form>

      <hr></hr>

      <h1>Register new user account</h1>
      <form onSubmit={handleSubmit}>
        <FormInput label={"Role: "} value={role} onChangeValue={setRole} />

        <FormInput
          label={"Username: "}
          value={username}
          onChangeValue={setUsername}
        />
        <FormInput
          label={"Full name: "}
          value={fullName}
          onChangeValue={setFullName}
        />
        <label>
          <strong>Password: </strong>
          <input
            type="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
          />
        </label>

        <button>Create user</button>
      </form>
    </div>
  );
}

export function RegisterNewAccountForTest({ registerApi }) {
  const [role, setRole] = useState("");
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [fullName, setFullName] = useState("");
  const navigate = useNavigate();

  function handleSubmit(e) {
    e.preventDefault();
    navigate("../");
  }

  registerApi.addUser({ role, username, fullName, password }); //For test

  function handleSubmitBack() {
    navigate("/");
  }

  return (
    <div>
      <form onSubmit={handleSubmitBack}>
        <button>Go back to start page</button>
      </form>

      <hr></hr>

      <h1>Register new user account</h1>
      <form onSubmit={handleSubmit}>
        <FormInput label={"Role: "} value={role} onChangeValue={setRole} />
        <FormInput
          label={"Username: "}
          value={username}
          onChangeValue={setUsername}
        />
        <FormInput
          label={"Full name: "}
          value={fullName}
          onChangeValue={setFullName}
        />
        <FormInput
          label={"Password: "}
          value={password}
          onChangeValue={setPassword}
        />
        <button>Log in</button>
      </form>
    </div>
  );
}
