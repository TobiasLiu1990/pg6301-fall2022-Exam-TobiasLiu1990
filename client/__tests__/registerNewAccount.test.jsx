import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import {RegisterNewAccount, RegisterNewAccountForTest} from "../registerNewAccount";
import { act, Simulate } from "react-dom/test-utils";

describe("register new account component", () => {
  it("should show register form", async () => {
    const element = document.createElement("div");

    await act(async () =>
      ReactDOM.render(
        <MemoryRouter>
          <RegisterNewAccount />
        </MemoryRouter>,
        element
      )
    );

    expect(element.innerHTML).toMatchSnapshot();

    const inputLabels = Array.from(
      element.querySelectorAll("form label strong")
    ).map((label) => label.innerHTML);

    expect(inputLabels).toEqual(["Username: ", "Full name: ", "Password: "]);
  });

  it("should create new account", async () => {
    const addUser = jest.fn();
    const element = document.createElement("div");

    await act(async () =>
      ReactDOM.render(
        <MemoryRouter>
          <RegisterNewAccountForTest registerApi={{addUser}} />
        </MemoryRouter>,
        element
      )
    );

    Simulate.change(
      element.querySelector("form:nth-of-type(2) div:nth-of-type(1) input"),
      {
        target: { value: "simulated user" },
      }
    );

    Simulate.change(
      element.querySelector("form:nth-of-type(2) div:nth-of-type(2) input"),
      {
        target: { value: "simulated full name" },
      }
    );

    Simulate.change(
      element.querySelector("form:nth-of-type(2) div:nth-of-type(3) input"),
      {
        target: { value: "simulated password" },
      }
    );

    Simulate.submit(element.querySelector("form"));

    expect(addUser).toBeCalledWith({
      username: "simulated user",
      fullName: "simulated full name",
      password: "simulated password",
    });
  });

  it("should go back on back button", async () => {
    const backOnClick = jest.fn();
    const element = document.createElement("div");


    // await act(async () =>
    //     ReactDOM.render(
    //         <MemoryRouter>
    //           <RegisterNewAccount handleClick={backOnClick} />
    //         </MemoryRouter>,
    //         element
    //     )
    // );
    //
    // Simulate.click(element.querySelector("button:nth-of-type(1)"))
    //
    // expect(backOnClick).toBeCalled();



    // const backOnClick = jest.fn();
    // const element = document.createElement("div");
    // const mockButton = ReactDOM.render((<button onClick={backOnClick}>Go Back</button>), element)
    // Simulate.click(element.querySelector("button"))
    // expect(backOnClick.mock.calls.length).toEqual(1)


  })
});
