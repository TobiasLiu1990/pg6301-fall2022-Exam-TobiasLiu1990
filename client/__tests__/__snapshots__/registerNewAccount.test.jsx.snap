// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register new account component should show register form 1`] = `"<div><form><button>Go back to start page</button></form><hr><h1>Register new user account</h1><form><div><label><strong>Role: </strong> <input value=""></label><br></div><div><label><strong>Username: </strong> <input value=""></label><br></div><div><label><strong>Full name: </strong> <input value=""></label><br></div><label><strong>Password: </strong><input type="password" value=""></label><button>Create user</button></form></div>"`;
